src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs

```diff
@@ -80,9 +185,41 @@ private static void ThrowLockTakenException()
         =========================================================================*/
         public static bool TryEnter(object obj)
         {
-            bool lockTaken = false;
-            TryEnter(obj, 0, ref lockTaken);
-            return lockTaken;
+            ArgumentNullException.ThrowIfNull(obj);
+
+            EnterHelperResult tryEnterResult = TryEnter_FastPath_WithTimeout(obj, 0);
+            if (tryEnterResult == EnterHelperResult.Entered)
+            {
+                return true;
+            }
+            else if (tryEnterResult == EnterHelperResult.Contention)
+            {
+                return false;
+            }
+
+            return TryEnter_Slowpath(obj);
+        }
+
+        private static void TryEnter_Timeout_WithLockTaken(object obj, int timeout, ref bool lockTaken)
+        {
+            if (timeout >= -1)
+            {
+                EnterHelperResult tryEnterResult = TryEnter_FastPath_WithTimeout(obj, timeout);
+                if (tryEnterResult == EnterHelperResult.Entered)
+                {
+                    lockTaken = true;
+                    return;
+                }
+                else if (tryEnterResult == EnterHelperResult.Contention)
+                {
+                    return;
+                }
+            }
+
+            if (TryEnter_Slowpath(obj, timeout))
```

Comment:
Will this result in _up to_ two `timeout` periods elapsing if the fast-path tried and returned anything other than Entered or Contention?

Reply:
No, the timeout input to the fast path is only used to check against 0 (which has special meaning).